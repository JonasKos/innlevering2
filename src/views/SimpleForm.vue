<template>
  <div>
    <h1>Leave feedback</h1>
    <form @submit.prevent="sendForm">
      <h3>Leave feedback for the calculator</h3>
      <div class="former">
        <!-- Name -->
        <div
          class="form-group"
          :class="{ error: v$.feedback.name.$errors.length }"
        >
          <label for="">Name</label>
          <input
            class="form-control"
            placeholder="Enter your name"
            type="text"
            v-model="v$.feedback.name.$model"
          />

          <!-- name error message -->
          <div class="input-container">
            <div
              class="input-errors"
              v-for="(error, index) of v$.feedback.name.$errors"
              :key="index"
            >
              <div class="error-msg">{{ error.$message }}</div>
            </div>
          </div>
        </div>

        <!-- Email -->
        <div
          class="form-group"
          :class="{ error: v$.feedback.email.$errors.length }"
        >
          <label for="">Email</label>
          <input
            class="form-control"
            placeholder="Enter your email"
            type="email"
            v-model="v$.feedback.email.$model"
          />

          <!-- email error message -->
          <div
            class="input-errors"
            v-for="(error, index) of v$.feedback.email.$errors"
            :key="index"
          >
            <div class="error-msg">{{ error.$message }}</div>
          </div>
        </div>
        <!-- feedback -->
        <div
          class="form-group"
          :class="{ error: v$.feedback.message.$errors.length }"
        >
          <label for="">Message</label>
          <input
            class="form-control"
            placeholder="Enter your feedback"
            type="text"
            v-model="v$.feedback.message.$model"
          />
          <!-- feedback error message -->
          <div
            class="input-errors"
            v-for="(error, index) of v$.feedback.message.$errors"
            :key="index"
          >
            <div class="error-msg">{{ error.$message }}</div>
          </div>
        </div>
      </div>

      <div>
        <button
          type="submit"
          :disabled="v$.feedback.$invalid"
          class="submit-button"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</template>

<script>
import axios from "axios";
import useVuelidate from "@vuelidate/core";
import { required, email, minLength } from "@vuelidate/validators";
import store from "../store";

export function validName(name) {
  let validNamePattern = new RegExp("^[a-zA-Z]+(?:[-'\\s][a-zA-Z]+)*$");
  if (validNamePattern.test(name)) {
    return true;
  }
  return false;
}

export default {
  name: "SimpleForm",

  setup() {
    return { v$: useVuelidate() };
  },
  
  data() {
    return {
      feedback: {
        name: store.state.name,
        email: store.state.email,
        message: "",
      },
    };
  },

  validations() {
    return {
      feedback: {
        name: {
          required,
          minLength: minLength(3),
          name_validation: {
            $validator: validName,
            $message:
              "Please enter a valid name. No special characters, numbers or spaces at the start or the end of the name are allowed.",
          },
        },
        email: {
          required,
          email,
        },
        message: {
          required,
          minLength: minLength(30),
        },
      },
    };
  },

  methods: {
    sendForm() {
      store.commit("setName", this.feedback.name);
      store.commit("setEmail", this.feedback.email);

      console.log(store.state.name + ", store.name");
      console.log(store.state.email + ", store.email");

      axios
        .post(
          "http://localhost:3000/feedback",
          console.log(this.feedback),
          this.feedback
        )
        .then(function (response) {
          alert("Form successfully submitted, Status code: " + response.status);
          console.log("response");
          console.log(response);
        })
        .catch(function (err) {
          alert(err);
          console.log(err);
        });
    },
  },
};
</script>

<style>

.input-errors {
  color: red;
}

.form-group {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  width: 100%;
  margin-bottom: 20px;
}
</style>
