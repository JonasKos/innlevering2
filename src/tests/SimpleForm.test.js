import { mount } from "@vue/test-utils";
import { describe, expect } from "vitest";
import SimpleForm from "../views/SimpleForm.vue";

describe("SimpleForm", () => {
  it("Invalid name", async () => {
    const wrapper = mount(SimpleForm);
    const nameInput = wrapper.find('[placeholder="Enter your name"]');
    await nameInput.setValue("testNameWithSpaceAtEnd ");
    expect(wrapper.find('[class="form-group error"]').exists()).toBe(true);
  });
  it("Valid name", async () => {
    const wrapper = mount(SimpleForm);
    const nameInput = wrapper.find('[placeholder="Enter your name"]');
    await nameInput.setValue("testName");
    expect(wrapper.find('[class="form-group error"]').exists()).toBe(false);
  });
  it("Invalid email", async () => {
    const wrapper = mount(SimpleForm);
    const emailInput = wrapper.find('[placeholder="Enter your email"]');
    await emailInput.setValue("InvalidEmail");
    expect(wrapper.find('[class="form-group error"]').exists()).toBe(true);
  });
  it("Valid email", async () => {
    const wrapper = mount(SimpleForm);
    const emailInput = wrapper.find('[placeholder="Enter your email"]');
    await emailInput.setValue("validEmail@valid.com");
    expect(wrapper.find('[class="form-group error"]').exists()).toBe(false);
  });
  it("Invalid message", async () => {

    const wrapper = mount(SimpleForm);
    const messageInput = wrapper.find('[placeholder="Enter your feedback"]');
    await messageInput.setValue("InvalidMessage");
    expect(wrapper.find('[class="form-group error"]').exists()).toBe(true);
  });
  it("Valid message", async () => {
    const wrapper = mount(SimpleForm);
    const messageInput = wrapper.find('[placeholder="Enter your feedback"]');
    await messageInput.setValue("ValidMessageValidMessageValidMessage");
    expect(wrapper.find('[class="form-group error"]').exists()).toBe(false);
  });
  it("Submit button disabled", async () => {
    const wrapper = mount(SimpleForm);
    const submitButton = wrapper.find('[type="submit"]');
    expect(submitButton.element.disabled).toBe(true);
  });
  it("Submit button enabled", async () => {
    const wrapper = mount(SimpleForm);
    const nameInput = wrapper.find('[placeholder="Enter your name"]');
    const emailInput = wrapper.find('[placeholder="Enter your email"]');
    const messageInput = wrapper.find('[placeholder="Enter your feedback"]');
    await nameInput.setValue("testName");
    await emailInput.setValue("validEmail@valid.com");
    await messageInput.setValue("ValidMessageValidMessageValidMessage");
    const submitButton = wrapper.find('[type="submit"]');
    expect(submitButton.element.disabled).toBe(false);
  });
});