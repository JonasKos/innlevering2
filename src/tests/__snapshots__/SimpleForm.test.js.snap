// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SimpleForm > Email invalid 1`] = `
"<div>
  <h1>Leave feedback</h1>
  <form>
    <h3>Leave feedback for the calculator</h3>
    <div class=\\"former\\">
      <!-- Name -->
      <div class=\\"form-group\\"><label for=\\"\\">Name</label><input class=\\"form-control\\" placeholder=\\"Enter your name\\" type=\\"text\\"><!-- name error message -->
        <div class=\\"input-container\\"></div>
      </div><!-- Email -->
      <div class=\\"form-group\\"><label for=\\"\\">Email</label><input class=\\"form-control\\" placeholder=\\"Enter your email\\" type=\\"email\\"><!-- email error message -->
      </div><!-- feedback -->
      <div class=\\"form-group\\"><label for=\\"\\">Message</label><input class=\\"form-control\\" placeholder=\\"Enter your feedback\\" type=\\"text\\"><!-- feedback error message -->
      </div>
    </div>
    <div><button type=\\"submit\\" disabled=\\"\\" class=\\"submit-button\\"> Submit </button></div>
  </form>
</div>"
`;

exports[`SimpleForm > renders a message 1`] = `
"<div>
  <h1>Leave feedback</h1>
  <form>
    <h3>Leave feedback for the calculator</h3>
    <div class=\\"former\\">
      <!-- Name -->
      <div class=\\"form-group\\"><label for=\\"\\">Name</label><input class=\\"form-control\\" placeholder=\\"Enter your name\\" type=\\"text\\"><!-- name error message -->
        <div class=\\"input-container\\"></div>
      </div><!-- Email -->
      <div class=\\"form-group\\"><label for=\\"\\">Email</label><input class=\\"form-control\\" placeholder=\\"Enter your email\\" type=\\"email\\"><!-- email error message -->
      </div><!-- feedback -->
      <div class=\\"form-group\\"><label for=\\"\\">Message</label><input class=\\"form-control\\" placeholder=\\"Enter your feedback\\" type=\\"text\\"><!-- feedback error message -->
      </div>
    </div>
    <div><button type=\\"submit\\" disabled=\\"\\" class=\\"submit-button\\"> Submit </button></div>
  </form>
</div>"
`;
